source/opt/inline_pass.cpp

```diff
@@ -270,30 +319,120 @@ std::unique_ptr<BasicBlock> InlinePass::AddGuardBlock(
   return new_blk_ptr;
 }
 
+DebugScope InlinePass::BuildDebugScope(const DebugScope& callee_instr_scope,
+                                       DebugInlinedAtContext& inlined_at_ctx) {
+  return DebugScope(callee_instr_scope.GetLexicalScope(),
+                    BuildDebugInlinedAtChain(callee_instr_scope.GetInlinedAt(),
+                                             inlined_at_ctx));
+}
+
+uint32_t InlinePass::BuildDebugInlinedAtChain(
+    uint32_t callee_inlined_at, DebugInlinedAtContext& inlined_at_ctx) {
+  if (inlined_at_ctx.FunctionCallScope().GetLexicalScope() == kNoDebugScope)
+    return kNoInlinedAt;
+
+  // Reuse the already generated DebugInlinedAt chain if exists.
+  uint32_t already_generated_chain_head_id =
+      inlined_at_ctx.InlinedAtChainHead(callee_inlined_at);
+  if (already_generated_chain_head_id != kNoInlinedAt) {
+    return already_generated_chain_head_id;
+  }
+
+  const uint32_t inlined_to_caller =
+      context()->get_debug_info_mgr()->CreateDebugInlinedAt(
+          inlined_at_ctx.FunctionCallLine(),
+          inlined_at_ctx.FunctionCallScope());
+  if (inlined_to_caller == kNoInlinedAt) return kNoInlinedAt;
+
+  // Simply specify that this instruction was inlined to the caller
+  // if it has no DebugInlinedAt in advance.
+  if (callee_inlined_at == kNoInlinedAt) {
+    // Keep the new chain information that will be reused it.
+    inlined_at_ctx.UpdateInlinedAtContext(kNoInlinedAt, inlined_to_caller);
+    return inlined_to_caller;
+  }
+
+  // Create a new recursive DebugInlinedAt chain.
+  uint32_t chain_head_id = kNoInlinedAt;
+  uint32_t chain_iter_id = callee_inlined_at;
+  Instruction* last_inlined_at_in_chain = nullptr;
+  do {
+    Instruction* new_inlined_at_in_chain =
+        context()->get_debug_info_mgr()->CloneDebugInlinedAt(
+            chain_iter_id, /* insert_before */ last_inlined_at_in_chain);
+
+    // Set DebugInlinedAt of the new scope as the head of the chain.
+    if (chain_head_id == kNoInlinedAt)
+      chain_head_id = new_inlined_at_in_chain->result_id();
+
+    // Previous DebugInlinedAt of the chain must point to the new
+    // DebugInlinedAt as its Inlined operand to build a recursive
+    // chain.
+    if (last_inlined_at_in_chain != nullptr) {
+      UpdateInlinedOperand(last_inlined_at_in_chain,
+                           new_inlined_at_in_chain->result_id());
+    }
+    last_inlined_at_in_chain = new_inlined_at_in_chain;
+
+    if (!HasInlinedOperand(new_inlined_at_in_chain)) {
+      break;
+    }
+    chain_iter_id = GetInlinedOperand(new_inlined_at_in_chain);
```

Comment:
Would be a better name `GetInlinedAtOperand`.  Also, why not include the check if the operand is present in this function?  It can return 0 if it does not exist.